<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="panda">

  <xacro:include filename="$(find franka_description)/robots/common/franka_arm.xacro"/>
  <xacro:include filename="$(find franka_description)/robots/common/franka_hand.xacro"/>
  
  <xacro:arg name="arm_id" default="panda" />
  <link name="world" />

  <xacro:franka_arm arm_id="$(arg arm_id)" connected_to="world" xyz="0 0 0" 
                    safety_distance="0.03" gazebo="true" 
                    joint_limits="${xacro.load_yaml('$(find franka_description)/robots/panda/joint_limits.yaml')}"/>

  <xacro:franka_hand arm_id="$(arg arm_id)" rpy="0 0 ${-pi/4}" 
                     connected_to="$(arg arm_id)_link8" safety_distance="0.03" gazebo="true"/>

  <link name="panda_TCP"/>
  <joint name="panda_TCP_joint" type="fixed">
    <parent link="panda_hand"/>
    <child link="panda_TCP"/>
    <origin xyz="0 0 0.1034" rpy="0 0 0"/>
  </joint>

  <link name="fork_tip">
    <visual>
      <origin xyz="-0.033 -0.02 0.0171378" rpy="0 ${27.5 * pi / 180} 0"/>
      <geometry><mesh filename="package://vision_processing/src/vision_processing/diffusion_model_train/fork.STL" scale="0.001 0.001 0.001"/></geometry>
    </visual>
    <!-- <collision>
      <origin xyz="-0.033 -0.02 0.0171378" rpy="0 ${27.5 * pi / 180} 0"/>
      <geometry><mesh filename="package://vision_processing/src/vision_processing/diffusion_model_train/fork.STL" scale="0.001 0.001 0.001"/></geometry>
    </collision> -->
    <inertial> <mass value="0.1" />
      <inertia ixx="0.0001" ixy="0" ixz="0" iyy="0.0001" iyz="0" izz="0.0001" />
    </inertial>
  </link>

  <joint name="fork_tip_joint" type="fixed">
    <origin rpy="0 ${-(180+27.5)/360*2*pi} 0" xyz="-0.0055 0 ${0.233-0.1034}"/>
    <parent link="panda_TCP"/>
    <child link="fork_tip"/>
  </joint>

  <xacro:macro name="create_transmission" params="joint_name">
    <transmission name="${joint_name}_tran">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="${joint_name}">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      </joint>
      <actuator name="${joint_name}_motor">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
    </transmission>
  </xacro:macro>

  <xacro:create_transmission joint_name="$(arg arm_id)_joint1"/>
  <xacro:create_transmission joint_name="$(arg arm_id)_joint2"/>
  <xacro:create_transmission joint_name="$(arg arm_id)_joint3"/>
  <xacro:create_transmission joint_name="$(arg arm_id)_joint4"/>
  <xacro:create_transmission joint_name="$(arg arm_id)_joint5"/>
  <xacro:create_transmission joint_name="$(arg arm_id)_joint6"/>
  <xacro:create_transmission joint_name="$(arg arm_id)_joint7"/>

  <joint name="camera_static_joint" type="fixed">
    <parent link="world" />
    <child link="camera_static_link" />
    <origin xyz="1.1593 1.2922 1.1910" rpy="0.0000 0.6013 -2.3020" />
    <!-- <origin xyz="1.1593 1.2922 1.1910" rpy="-1.967 0.009 2.559" /> -->
  </joint>

  <link name="camera_static_link">
    <visual>
      <geometry><box size="0.025 0.09 0.02"/></geometry>
    </visual>
    <inertial>
      <mass value="0.1" /><inertia ixx="1e-4" ixy="0" ixz="0" iyy="1e-4" iyz="0" izz="1e-4" />
    </inertial>
  </link>

  <joint name="camera_static_optical_joint" type="fixed">
    <parent link="camera_static_link"/><child link="camera_static_optical_frame"/>
    <origin xyz="0 0 0" rpy="${-pi/2} 0 ${-pi/2}"/>
  </joint>
  <link name="camera_static_optical_frame"/>

  <joint name="camera_wrist_joint" type="fixed">
    <parent link="panda_TCP"/> 
    <child link="camera_wrist_link"/>
    <origin xyz="-0.052 0.035 -0.045" rpy="0 ${-pi/2} 0"/> 
  </joint>

  <link name="camera_wrist_link">
    <visual>
      <geometry><box size="0.025 0.09 0.02"/></geometry>
    </visual>
    <inertial>
      <mass value="0.1" /><inertia ixx="1e-4" ixy="0" ixz="0" iyy="1e-4" iyz="0" izz="1e-4" />
    </inertial>
  </link>

  <joint name="camera_wrist_optical_joint" type="fixed">
    <parent link="camera_wrist_link"/><child link="camera_wrist_optical_frame"/>
    <origin xyz="0 0 0" rpy="${-pi/2} 0 ${-pi/2}"/>
  </joint>
  <link name="camera_wrist_optical_frame"/>

  <gazebo reference="camera_static_link">
    <material>Gazebo/Red</material>
    <sensor type="depth" name="camera_static">
      <update_rate>6.0</update_rate>
      <camera>
        <horizontal_fov>0.958</horizontal_fov> 
        <image><width>640</width><height>480</height><format>R8G8B8</format></image>
        <clip><near>0.1</near><far>10.0</far></clip>
      </camera>
      <plugin name="static_camera_controller" filename="libgazebo_ros_openni_kinect.so">
        <cameraName>camera_static</cameraName>
        <alwaysOn>true</alwaysOn>
        <imageTopicName>color/image_raw</imageTopicName>
        <depthImageTopicName>aligned_depth_to_color/image_raw</depthImageTopicName>
        <cameraInfoTopicName>color/camera_info</cameraInfoTopicName>
        <frameName>camera_static_optical_frame</frameName>
      </plugin>
    </sensor>
  </gazebo>

  <gazebo reference="camera_wrist_link">
    <material>Gazebo/Blue</material>
    <sensor type="depth" name="camera_wrist">
      <update_rate>6.0</update_rate>
      
      <camera>
        <horizontal_fov>0.973</horizontal_fov>
        <image>
            <width>640</width>
            <height>480</height>
            <format>R8G8B8</format>
        </image>
        <clip>
            <near>0.01</near> <far>10.0</far>
        </clip>
      </camera>

      <plugin name="wrist_camera_controller" filename="libgazebo_ros_openni_kinect.so">
        <cameraName>camera_wrist</cameraName>
        <alwaysOn>true</alwaysOn>
        <imageTopicName>color/image_raw</imageTopicName>
        <depthImageTopicName>aligned_depth_to_color/image_raw</depthImageTopicName>
        <cameraInfoTopicName>color/camera_info</cameraInfoTopicName>
        <frameName>camera_wrist_optical_frame</frameName>
        
        <pointCloudCutoff>0.01</pointCloudCutoff>
        <pointCloudCutoffMax>10.0</pointCloudCutoffMax>
        
        <updateRate>6.0</updateRate>
      </plugin>
    </sensor>
  </gazebo>
  
  <gazebo>
    <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
      <robotNamespace>/</robotNamespace>
      <robotSimType>gazebo_ros_control/DefaultRobotHWSim</robotSimType>
    </plugin>
  </gazebo>
</robot>